### PR TITLE
Prune stale issues

### LINKED ISSUE #1762: Block indentation in chains
Shouldn't chains with visual indentation indent blocks to the same visual indentation?

```rust
let a = b.iter()
         .map(|m| {
             m.hi()
         });
```

Instead of

```rust
let a = b.iter()
         .map(|m| {
             m.hi()
});
```

### LINKED ISSUE #2388: Format macro def with repeat
```
thread 'main' panicked at 'Box<Any>', /Users/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-syntax-12.0.0/parse/mod.rs:228:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: syntax::parse::filemap_to_stream
   7: syntax::parse::filemap_to_parser
   8: syntax::parse::new_parser_from_source_str
   9: rustfmt_nightly::parse_input
  10: rustfmt_nightly::format_snippet
  11: rustfmt_nightly::macros::rewrite_macro_def
  12: rustfmt_nightly::visitor::FmtVisitor::visit_item
  13: rustfmt_nightly::visitor::FmtVisitor::walk_items
  14: rustfmt_nightly::visitor::FmtVisitor::walk_mod_items
  15: rustfmt_nightly::visitor::FmtVisitor::format_separate_mod
  16: rustfmt_nightly::run
  17: rustfmt::execute
  18: rustfmt::main
  19: std::rt::lang_start::{{closure}}
  20: std::panicking::try::do_call
  21: __rust_maybe_catch_panic
  22: std::rt::lang_start_internal
  23: main
```

Just running `cargo fmt` at the root of nrc/graphql. Also happens at nrc/graphql/graphql. I have no idea why.

### LINKED ISSUE #2672: No indentation for functions exceeding line maximum width if overflow can't be handled without exceeding line width
Much easier to show with an example than explain:

Before fmt:
```
fn main() {
    fn test(averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32){}
    fn test(shortname : i32, averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32){}

    let averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff = 0;
}

mod asdf {
    fn test(averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32){}
}
```

After fmt:
```
fn main() {
fn test(averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32){
    }
fn test(shortname : i32, averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32){
    }

    let averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff =
        0;
}

mod asdf {
fn test(averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32){
    }
}
```

This is probably a degenerative case, but the similar assignment with a long variable/type seems to be handled better. In practice I noticed this when rightward drift from closures got out of hand.

### LINKED ISSUE #2755: Macro matchers inconsistently having whitespace removed after `,`
It seems that rustfmt is sometimes trying to remove whitespace after `,` when it appears in macro matchers, but not always. Diff from Diesel:

```patch
--- a/diesel/src/expression/operators.rs
+++ b/diesel/src/expression/operators.rs
@@ -106,30 +106,18 @@ macro_rules! __diesel_operator_body {
 #[macro_export]
 #[doc(hidden)]
 macro_rules! __diesel_operator_to_sql {
-    (
-        notation = infix,
-        operator_expr = $op:expr,
-        field_exprs = ($left:expr, $right:expr),
-    ) => {
+    (notation = infix,operator_expr = $op:expr,field_exprs = ($left:expr, $right:expr),) => {
         $left;
         $op;
         $right;
     };

-    (
-        notation = postfix,
-        operator_expr = $op:expr,
-        field_exprs = ($expr:expr),
-    ) => {
+    (notation = postfix,operator_expr = $op:expr,field_exprs = ($expr:expr),) => {
         $expr;
         $op;
     };

-    (
-        notation = prefix,
-        operator_expr = $op:expr,
-        field_exprs = ($expr:expr),
-    ) => {
+    (notation = prefix,operator_expr = $op:expr,field_exprs = ($expr:expr),) => {
         $op;
         $expr;
     };
@@ -258,7 +246,7 @@ macro_rules! diesel_postfix_operator {
         diesel_postfix_operator!($name, $operator, $crate::sql_types::Bool);
     };

-    ($name:ident, $operator:expr, backend: $backend:ty) => {
+    ($name:ident, $operator:expr,backend: $backend:ty) => {
         diesel_postfix_operator!($name, $operator, $crate::sql_types::Bool, backend: $backend);
     };

@@ -275,7 +263,7 @@ macro_rules! diesel_postfix_operator {
         );
     };

-    ($name:ident, $operator:expr, $return_ty:ty, backend: $backend:ty) => {
+    ($name:ident, $operator:expr, $return_ty:ty,backend: $backend:ty) => {
         __diesel_operator_body!(
             notation = postfix,
             struct_name = $name,
@@ -303,7 +291,7 @@ macro_rules! diesel_prefix_operator {
         diesel_prefix_operator!($name, $operator, $crate::sql_types::Bool);
     };

-    ($name:ident, $operator:expr, backend: $backend:ty) => {
+    ($name:ident, $operator:expr,backend: $backend:ty) => {
         diesel_prefix_operator!($name, $operator, $crate::sql_types::Bool, backend: $backend);
```

```patch
--- a/diesel/src/macros/mod.rs
+++ b/diesel/src/macros/mod.rs
@@ -839,7 +839,7 @@ macro_rules! __diesel_table_impl {
 #[macro_export]
 #[doc(hidden)]
 macro_rules! __diesel_table_query_source_impl {
-    ($table_struct:ident, public, $table_name:expr) => {
+    ($table_struct:ident,public, $table_name:expr) => {
         impl QuerySource for $table_struct {
             type FromClause = Identifier<'static>;
             type DefaultSelection = <Self as Table>::AllColumns;
@@ -966,12 +966,13 @@ macro_rules! joinable_inner {
     };

     (
-        left_table_ty = $left_table_ty:ty,
-        right_table_ty = $right_table_ty:ty,
-        right_table_expr = $right_table_expr:expr,
-        foreign_key = $foreign_key:path,
-        primary_key_ty = $primary_key_ty:ty,
-        primary_key_expr = $primary_key_expr:expr,
+        left_table_ty =
+        $left_table_ty:ty,right_table_ty =
+        $right_table_ty:ty,right_table_expr =
+        $right_table_expr:expr,foreign_key =
+        $foreign_key:path,primary_key_ty =
+        $primary_key_ty:ty,primary_key_expr =
+        $primary_key_expr:expr,
     ) => {
         impl $crate::JoinTo<$right_table_ty> for $left_table_ty {
             type FromClause = $right_table_ty;
```

### LINKED ISSUE #2947: Bad formatting of `write!` as match arm
[playground](https://play.rust-lang.org/?gist=a2fe58ef80175832a64c186457f996ae&version=nightly&mode=debug&edition=2015)

```rust
#[derive(Copy, Clone, Debug)]
enum DeriveError {
    UnionUnsupported(String),
    UnitStructUnsupported(String),
    NamedStructLacksSpan(String),
    TupleStructNotNewtype(String),
}

impl fmt::Display for DeriveError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DeriveError::UnionUnsupported(name) => {
                write!(f, "Cannot derive `Spaned` for `union {}`", name)
            }
            DeriveError::UnitStructUnsupported(name) => {
                write!(f, "Cannot derive `Spanned` for `struct {};`", name)
            }
            DeriveError::NamedStructLacksSpan(name) => write!(
                f,
                "Cannot derive `Spanned` for `struct {}` as it lacks a field `span`",
                name
            ),
            DeriveError::TupleStructNotNewtype(name) => {
                write!(f,
                "Cannot derive `Spanned` for `struct {}` as it does not have a single tuple member",
                name,
            )
            }
        }
    }
}
```

This is stable. Adding a newline after the `(` in `write!(f,` causes it to format correctly as the above arm. Removing the whitespace between `(` and `f` _in the last arm only_ results in this bad formatting again.

(Yes I know the error is not `Copy` this was during editing that I noticed the bad formatting...)

### LINKED ISSUE #2978: New line inserted between ) and { in ClosingNextLine mode when line length is 99
rustfmt formats inserts a newline between `)` and `{` in this example:

```
pub fn test() {
    let xxxxxxxxxxxxxxxxxxxxxxxxxxxx = yyyyyyyyyyyyyyyy | zzzzzzzzzzzzzzzzzzzzzzzz | if for_writing
    {
        fffffffffffffffffffff
    }
    else {
        g
    };
}
```

When the `let xxx…` line has 99 characters.

I would expect a formatting like this (which rustfmt produces when adding one more `x`):

```
pub fn test() {
    let xxxxxxxxxxxxxxxxxxxxxxxxxxxxx =
        yyyyyyyyyyyyyyyy | zzzzzzzzzzzzzzzzzzzzzzzz | if for_writing {
            fffffffffffffffffffff
        }
        else {
            g
        };
}
```

### LINKED ISSUE #3148: normalize_comments closing brace indentation bug with hard_tabs
Hi,

with the following config:
```toml
hard_tabs = true
normalize_comments = true
```

the snippet:
```rust
/// ```
/// Data {
///     a: "some text data",
///     ..Default::default()
/// };
/// ```
#[derive(Default)]
pub struct Data {
	a: &str,
	b: u32,
}
```

gets formatted as:
```rust
/// ```
/// Data {
/// 	a: "some text data",
/// 	..Default::default()
/// 	};
/// ```
#[derive(Default)]
pub struct Data {
	a: &str,
	b: u32,
}
```

Please note the (wrongly) indented closing brace in the doc code block.  With `hard_tabs = false` it works fine.

Tested with `rustfmt 0.99.6-nightly (750b252 2018-10-18)`.

I'd also appreciate it if there was a separate `comment_hard_tabs` setting (or it simply would always be `false`), as I think tabs should only be at the beginning of a line, and not within comments - if you're open to it I can open another issue.

### LINKED ISSUE #3206: rustfmt fails to format a specific match statement
I couldn't manage to isolate the specific bug in this case, so I'm going to provide the definition of the function containing the `match` statement as-is:

```rust
#[derive(Debug)]
enum Error<'a> {
    RuntimeError(&'a str),
    IoError(io::Error),
}

impl<'a> std::error::Error for Error<'a> {
    fn source(&self) -> Option<&(std::error::Error + 'static)> {
        match self {
            Error::IoError(error) => Some(error),
            _ => None,
        }
    }
}

impl<'a> Display for Error<'a> {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        match self {
            Error::RuntimeError(description) => write!(f, "runtime error: {}", description),
            Error::IoError(error) => write!(f, "I/O error: {}", error),
        }
    }
}

impl<'a> From<std::io::Error> for Error<'a> {
    fn from(error: io::Error) -> Self {
        Error::IoError(error)
    }
}

type Result<'a, T> = std::result::Result<T, Error<'a>>;

enum Value {
    Str(String),
    Bool(bool),
    Number(Float),
}

impl Display for Value {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        match self {
            Value::Str(value) => write!(f, "{}", value),
            Value::Bool(value) => write!(f, "{}", value),
            Value::Number(value) => write!(f, "{}", value),
        }
    }
}

enum InputMode {
    Disabled,
    Simple(String),
    Nested(usize, String),
}

struct State {
    stack: Vec<Value>,
    input_mode: InputMode,
}

enum ArithmeticOperation {
    Addition,
    Subtraction,
    Multiplication,
    Division,
    Remainder,
}

fn apply_arithmetic_operation<'a>(op: ArithmeticOperation, state: &mut State) -> Result<'a, ()> {
    match       (       state.stack.pop           (),       state.stack.pop      ()                   )





    {
                    (Some(value1), Some(value2)) => {



                        match (value1, value2) {
                        (Value::Number(num1), Value::Number(num2)) => {
                            let result = match op {
                                ArithmeticOperation::Addition => num1 + num2,
                                ArithmeticOperation::Subtraction => num1 - num2,
                                ArithmeticOperation::Multiplication => num1 * num2,
                                    ArithmeticOperation::Division => num1 /num2,
                                ArithmeticOperation::Remainder => num1 % num2,
                            };
                            state.stack.push(Value::Number(result));
                            Ok(())
                        }
                        _ =>
                    Err(Error::RuntimeError(
                        "cannot sum values on top of stack: the two topmost values on the stack should be numbers",
                    ))
                    }}
                    _ =>
                                                return Err(Error::RuntimeError(
                        "cannot sum values on top of stack: there must be at least two values on the stack",
                    ))
                }
}
```

`rustfmt` seems to properly format that except for the whole outermost `match` statement where the whitespace still remains garbled.

### PR DESCRIPTION
**Pull request by ayazhafiz**
Adds regression tests for the following issues which seem to be fixed on
master:

Closes #1762
Closes #2388
Closes #2672
Closes #2755
Closes #2947
Closes #2978
Closes #3148
Closes #3206

@topecongiro @calebcartwright appologies for the large number of issues
in this commit; if you prefer I can split it up into 2+.

### CODE REVIEW

**Comment by calebcartwright** [comment]
No worries @ayazhafiz, for now this is fine as one PR

File: tests/target/issue-1762.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+fn main() {
+    let grammar: Vec<_> = ast
+        .attrs
+        .iter()
+        .filter(|attr| match attr.value {
+            MetaItem::NameValue(ref ident, _) => format!("{}", ident) == "grammar",
+            _ => false,
+        })
+        .collect();
+}
```
Comment (calebcartwright) [inline tests/target/issue-1762.rs:12]: Could we also include the snippet from the OP in the referenced iss
```rust
let a = b.iter()
         .map(|m| {
             m.hi()
});
```

File: tests/target/issue-2672.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    let averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff =
+        0;
+}
+
+mod asdf {
+    fn test(
+        averylongname_asdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdsffffffffffffffffffffffffffffffffffffffffffffffff: i32,
+    ) {
+    }
+}
```
Comment (calebcartwright) [inline tests/target/issue-2672.rs:21]: Could we include the other snippets posted in https://github.com/rust-lang/rustfmt/issues/2672#issuecomment-416498411 on that thread?

**Comment by ayazhafiz** [comment]
Yeah I'll add them in a couple hours

**Pull request merged**